### PR TITLE
Include file paths in tap runner output

### PR DESCRIPTION
**Pull request by snehaso**
This PR changes the test-unit-runner-tap to include source file location for each tests. This will be useful in calculating duration taken by each test file.

This is done by extending listener methods in the BaseTestRunner class. This makes it easy to be updated with the parent branch.

JSON output: https://gist.github.com/snehaso/6aedfdd18d91cb1486da

### CODE REVIEW

**Comment by darkofabijan** [comment]
@snehaso Looks good! @shiroyasha @strika Can you also take a look? Thanks!

File: lib/test/unit/ui/tap/ext/base_testrunner.rb [hunk 1]
```diff
@@ -0,0 +1,62 @@
+require 'test/unit/ui/tap/base_testrunner'
+
+module Test
+  module Unit
+    module UI
+      module Tap
+        class Tap::BaseTestRunner < Test::Unit::UI::TestRunner
```
Comment (shiroyasha) [inline lib/test/unit/ui/tap/ext/base_testrunner.rb]: The additional `Tap::` in the class name is confusing me a little. The class is already in the `Tap` module :)

Is it necessary?

File: lib/test/unit/ui/tap/ext/base_testrunner.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            doc = tapout_pass_without_source_location(test)
+            if(doc && doc['type'] == 'test')
+              file, _ = test.class.instance_method(clean_label(test.name).to_sym).source_location
+              doc['file'] = file.sub(Dir.pwd+'/', '')
+            end
+            doc
+          end
+
+          def tapout_todo_with_source_location(fault)
+            doc = tapout_todo(fault)
```
Comment (shiroyasha) [inline lib/test/unit/ui/tap/ext/base_testrunner.rb]: I am not 100% sure, but I think this line will cause an infinite recursion loop. Should it be rewritten like this maybe:

``` ruby
doc = tapout_todo_without_source_location(fault)
```

File: lib/test/unit/ui/tap/ext/base_testrunner.rb [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            doc['file'] = extract_source_location
```
Comment (shiroyasha) [inline lib/test/unit/ui/tap/ext/base_testrunner.rb]: `extract_source_location` takes `fault` as an argument I think

Comment (shiroyasha) [inline lib/test/unit/ui/tap/ext/base_testrunner.rb]: Also, in the other methods you are testing for the existence of the `doc` before adding the source location. Is it safe to add the file location to the doc without a check in this case?

**Comment by shiroyasha** [comment]
It looks very nice. I like your approach of modifying existing classes :clap: 

`tapout_todo_with_source_location` seems to be a bit broken thou. Perhaps we could introduce some additional tests, just to be 100% sure that it works?

File: lib/test/unit/ui/tap/ext/base_testrunner.rb (hunk 1 above)
Comment (snehaso) [inline lib/test/unit/ui/tap/ext/base_testrunner.rb]: Yes you're right. It is unnecessary.

File: lib/test/unit/ui/tap/ext/base_testrunner.rb (hunk 2 above)
Comment (snehaso) [inline lib/test/unit/ui/tap/ext/base_testrunner.rb]: Thanks for catching this. It was running in an infinite loop. 

I've fixed this in https://github.com/renderedtext/test-unit-runner-tap/commit/9454d2821bb26e0f9880188a1a6eeda5616d96a4 and have added a test.

File: lib/test/unit/ui/tap/ext/base_testrunner.rb (hunk 3 above)
Comment (snehaso) [inline lib/test/unit/ui/tap/ext/base_testrunner.rb]: @shiroyasha The `tapout_pass` method returns a nil if a test was already processed with a fault.  That means, it was already handled by one of the `tapout_(error| fail | todo| omit)` methods.  Here is the [code snippet](https://github.com/renderedtext/test-unit-runner-tap/blob/master/lib/test/unit/ui/tap/base_testrunner.rb#L159). 

The rest of the methods always return a `doc`, hence I did not add check for existence of `doc`.

**Comment by snehaso** [comment]
@shiroyasha This PR is ready for another review. I've fixed `tapout_todo_with_source_location` and added tests for todo and omitted scenarios.

**Comment by strika** [comment]
Looks great. :+1:

**Comment by shiroyasha** [comment]
@snehaso the new `tapout_todo_with_source_location` look great :+1:

**Pull request merged**